platform-sdk/docs/core/tipset-algorithm.md

```diff
@@ -0,0 +1,260 @@
+
+1) this is the first event a node has ever created (i.e. this is genesis), or
+2) the snapshot improvement score exceeds the snapshot improvement score of the previous event, or
+3) the snapshot improvement score is greater than zero and this is the first event created since
+   we last updated the snapshot
+
+By following this rule, this ensures that each node stops creating new events in a finite amount of time if that
+node no longer is in communicating with a quorum of its peers.
+
+```
+Example
+
+Address book weights: A = 5, B = 9, C = 11, D = 2
+Consider from A's perspective. A's snapshot advancement threshold is 14.
+
+Initially, the snapshot is [-1, -1, -1, -1].
+
+A creates a genesis event with tipset [0, -1, -1, -1]. The snapshot improvement score is 0, which is ok since
+this is a genesis event. (It will never again be ok to have an advancement score of 0.)
+
+Next, A creates an event with self parent [0, -1, -1, -1] and other parent [-1, 2, -1, -1] with generation 2.
+The tipset of the new event is [3, 2, -1, -1]. The partial weighted snapshot advancement score is 9.
+
+Next, A creates an event with self parent [3, 2, -1, -1] and other parent [3, 2, 5, 3] with generation 3. The tipset
+of the new event is [4, 2, 5, 3]. The partial weighted snapshot advancement score is 22, which exceeds the threshold
+of 14. Therefore, the snapshot is updated to [4, 2, 5, 3].
+
+When it next comes time for A to create an event, it will start comparing against this new snapshot [4, 2, 5, 3].
+```
+
+# Optimizing Tipset Score
+
+When it comes time to create a new event, if there are multiple other parents to choose from, select the other parent
+that results in the greatest tipset score improvement.
+
+# Bully Scores
+
+Informally, bullying is when the network decides (for whatever reason) not to use an event creator's events as other
+parents very often. The node being bullied (the nerd) will then find that it is difficult or impossible for its events
+to reach consensus.
+
+A bully score is a numeric value that increases the more a "nerd" is getting ignored. The more time that passes
+without putting a nerd's event into the ancestry of recent events, the higher the bully score is against that node.
+
+To compute the bully score against a node X, look at recent snapshot tipsets. Starting with the current snapshot
+and going backwards towards older snapshots, count the number of snapshots that need to be iterated over before an
+advancement in X's generation is observed. That count is the bully score against X.
+
+In order to prevent nodes from being bullied too badly, it is important to periodically choose a nerdy event as an
+other parent, even if that choice of other parent does not improve the snapshot advancement score much. The exact
+algorithm for this is a bit of a heuristic, and needs to be tuned on a variety of factors. Failure to occasionally
+select nerdy events may cause some nodes to be unable to create events that reach consensus.
```

Comment:
Is there a distribution of weighted probability of being selected and the distribution is the heuristic?  Or is there are hard threshold under which a specific node's event being selected is guaranteed?

Reply:
It's weighted proportionally to how much a node is being bullied. If you are being bullied twice as hard as another node, you are twice as likely to be chosen as a parent when we are running the anti-bullying algorithm.

I'd prefer not to get too deep in the weeds on the current heuristic in this particular document, since it's just a heuristic I chose that made tests look good, and is very likely to change in the future.